File: backend/app/services/unit_converter.py
from typing import Dict
from dataclasses import dataclass


@dataclass
class UnitDefinition:
    base_unit: str
    to_base: float
    from_base: float


class UnitConverter:
    """Converts between different unit systems"""

    # Base units: kg, mmol/L, mmHg, cm
    UNIT_CONVERSIONS: Dict[str, Dict[str, UnitDefinition]] = {
        "weight": {
            "kg": UnitDefinition("kg", 1.0, 1.0),
            "lbs": UnitDefinition("kg", 0.453592, 2.20462),
            "st": UnitDefinition("kg", 6.35029, 0.157473),
        },
        "height": {
            "cm": UnitDefinition("cm", 1.0, 1.0),
            "m": UnitDefinition("cm", 100.0, 0.01),
            "in": UnitDefinition("cm", 2.54, 0.393701),
            "ft": UnitDefinition("cm", 30.48, 0.0328084),
        },
        "glucose": {
            "mmol/L": UnitDefinition("mmol/L", 1.0, 1.0),
            "mg/dL": UnitDefinition("mmol/L", 0.05551, 18.0182),
        },
        "blood_pressure": {"mmHg": UnitDefinition("mmHg", 1.0, 1.0)},
    }

    def convert(
        self, value: float, from_unit: str, to_unit: str, category: str
    ) -> float:
        """Convert value from one unit to another"""
        if from_unit == to_unit:
            return value

        conversions = self.UNIT_CONVERSIONS.get(category)
        if not conversions:
            raise ValueError(f"Unknown unit category: {category}")

        from_def = conversions.get(from_unit)
        to_def = conversions.get(to_unit)

        if not from_def or not to_def:
            raise ValueError(f"Unsupported units: {from_unit}, {to_unit}")

        # Convert to base unit, then to target unit
        base_value = value * from_def.to_base
        return base_value * to_def.from_base

    def convert_to_user_unit(
        self, value: float, from_unit: str, user_unit: str, category: str
    ) -> float:
        """Convert value to user's preferred unit"""
        return self.convert(value, from_unit, user_unit, category)

    def convert_to_base_unit(
        self, value: float, from_unit: str, category: str
    ) -> float:
        """Convert value to base unit for storage"""
        conversions = self.UNIT_CONVERSIONS.get(category)
        if not conversions:
            raise ValueError(f"Unknown unit category: {category}")

        unit_def = conversions.get(from_unit)
        if not unit_def:
            raise ValueError(f"Unsupported unit: {from_unit}")

        return value * unit_def.to_base

    def get_supported_units(self, category: str) -> list:
        """Get list of supported units for a category"""
        conversions = self.UNIT_CONVERSIONS.get(category)
        if not conversions:
            return []
        return list(conversions.keys())

File: frontend/src/utils/unitConverter.ts
interface UnitDefinition {
  baseUnit: string;
  toBase: number;
  fromBase: number;
}

interface UnitConversions {
  [key: string]: {
    [key: string]: UnitDefinition;
  };
}

const UNIT_CONVERSIONS: UnitConversions = {
  weight: {
    kg: { baseUnit: 'kg', toBase: 1.0, fromBase: 1.0 },
    lbs: { baseUnit: 'kg', toBase: 0.453592, fromBase: 2.20462 },
    st: { baseUnit: 'kg', toBase: 6.35029, fromBase: 0.157473 },
  },
  height: {
    cm: { baseUnit: 'cm', toBase: 1.0, fromBase: 1.0 },
    m: { baseUnit: 'cm', toBase: 100.0, fromBase: 0.01 },
    in: { baseUnit: 'cm', toBase: 2.54, fromBase: 0.393701 },
    ft: { baseUnit: 'cm', toBase: 30.48, fromBase: 0.0328084 },
  },
  glucose: {
    'mmol/L': { baseUnit: 'mmol/L', toBase: 1.0, fromBase: 1.0 },
    'mg/dL': { baseUnit: 'mmol/L', toBase: 0.05551, fromBase: 18.0182 },
  },
  bloodPressure: {
    mmHg: { baseUnit: 'mmHg', toBase: 1.0, fromBase: 1.0 },
  },
};

export class UnitConverter {
  convert(value: number, fromUnit: string, toUnit: string, category: string): number {
    if (fromUnit === toUnit) {
      return value;
    }

    const conversions = UNIT_CONVERSIONS[category];
    if (!conversions) {
      throw new Error(`Unknown unit category: ${category}`);
    }

    const fromDef = conversions[fromUnit];
    const toDef = conversions[toUnit];

    if (!fromDef || !toDef) {
      throw new Error(`Unsupported units: ${fromUnit}, ${toUnit}`);
    }

    const baseValue = value * fromDef.toBase;
    return baseValue * toDef.fromBase;
  }

  convertToUserUnit(
    value: number,
    fromUnit: string,
    userUnit: string,
    category: string
  ): number {
    return this.convert(value, fromUnit, userUnit, category);
  }

  convertToBaseUnit(value: number, fromUnit: string, category: string): number {
    const conversions = UNIT_CONVERSIONS[category];
    if (!conversions) {
      throw new Error(`Unknown unit category: ${category}`);
    }

    const unitDef = conversions[fromUnit];
    if (!unitDef) {
      throw new Error(`Unsupported unit: ${fromUnit}`);
    }

    return value * unitDef.toBase;
  }
}

export const unitConverter = new UnitConverter();